core/src/main/java/org/apache/calcite/rel/rules/DateRangeRules.java

```diff
@@ -0,0 +1,361 @@
+ *
+ * <blockquote>SELECT ... FROM sales JOIN time_by_day USING (time_id)
+ * WHERE time_by_day.the_year = 1997
+ * AND time_by_day.the_month IN (4, 5, 6)</blockquote>
+ *
+ * <p>into
+ *
+ * <blockquote>SELECT ... FROM sales JOIN time_by_day USING (time_id)
+ * WHERE the_date BETWEEN DATE '2016-04-01' AND DATE '2016-06-30'</blockquote>
+ *
+ * <p>and is especially useful for Druid, which has a single timestamp column.
+ */
+public abstract class DateRangeRules {
+
+  private DateRangeRules() {}
+
+  private static final Predicate<Filter> FILTER_PREDICATE =
+      new Predicate<Filter>() {
+        @Override public boolean apply(Filter filter) {
+          final ExtractFinder finder = ExtractFinder.THREAD_INSTANCES.get();
+          assert finder.timeUnits.isEmpty() : "previous user did not clean up";
+          try {
+            filter.getCondition().accept(finder);
+            return !finder.timeUnits.isEmpty();
+          } finally {
+            finder.timeUnits.clear();
+          }
+        }
+      };
+
+  public static final RelOptRule FILTER_INSTANCE =
+      new FilterDateRangeRule(RelFactories.LOGICAL_BUILDER);
+
+  private static final Map<TimeUnitRange, Integer> TIME_UNIT_CODES =
+      ImmutableMap.<TimeUnitRange, Integer>builder()
+          .put(TimeUnitRange.YEAR, Calendar.YEAR)
+          .put(TimeUnitRange.MONTH, Calendar.MONTH)
+          .put(TimeUnitRange.DAY, Calendar.DAY_OF_MONTH)
+          .put(TimeUnitRange.HOUR, Calendar.HOUR)
+          .put(TimeUnitRange.MINUTE, Calendar.MINUTE)
+          .put(TimeUnitRange.SECOND, Calendar.SECOND)
+          .put(TimeUnitRange.MILLISECOND, Calendar.MILLISECOND)
+          .build();
+
+  private static final Map<TimeUnitRange, TimeUnitRange> TIME_UNIT_PARENTS =
+      ImmutableMap.<TimeUnitRange, TimeUnitRange>builder()
+          .put(TimeUnitRange.MONTH, TimeUnitRange.YEAR)
+          .put(TimeUnitRange.DAY, TimeUnitRange.MONTH)
+          .put(TimeUnitRange.HOUR, TimeUnitRange.DAY)
+          .put(TimeUnitRange.MINUTE, TimeUnitRange.HOUR)
+          .put(TimeUnitRange.SECOND, TimeUnitRange.MINUTE)
+          .put(TimeUnitRange.MILLISECOND, TimeUnitRange.SECOND)
+          .put(TimeUnitRange.MICROSECOND, TimeUnitRange.SECOND)
+          .build();
+
+  /** Returns whether an expression contains one or more calls to the
+   * {@code EXTRACT} function. */
+  @VisibleForTesting
+  static Set<TimeUnitRange> extractTimeUnits(RexNode e) {
+    final ExtractFinder finder = ExtractFinder.THREAD_INSTANCES.get();
+    try {
+      assert finder.timeUnits.isEmpty() : "previous user did not clean up";
+      e.accept(finder);
+      return ImmutableSet.copyOf(finder.timeUnits);
+    } finally {
+      finder.timeUnits.clear();
+    }
+  }
+
+  /** Rule that converts EXTRACT in a Filter condition into a date range. */
+  private static class FilterDateRangeRule extends RelOptRule {
+    FilterDateRangeRule(RelBuilderFactory relBuilderFactory) {
```

Comment:
I think we need to change visibility if we want to use it in Hive and plug our own builder factory.